# Fix downloaded dataset

### Redownload missing data

In [1]:
import os
import csv
import time
import requests

In [ ]:
fdir = '../../../../data/databases/Sabio-RK/'
fname = 'dataset_download.tsv'
f = open(fdir+fname, 'r')
lines = f.readlines()
f.close()

Certain database entries can be downloaded incompletely with the last line mostly missing (without `\n`) for some ungodly reason. Manually fixing these missing entries in the corresponding text files by redownloading the corresponding entries.

In [20]:
for line in lines:
    data = line.strip().split('\t')
    if len(data) != 16:
        print('EntryID:', data[0])
        print(line)

EntryID: 57731
57731	H2O;GM3	wildtype	7325957	Paenarthrobacter ureafaciens	Q5W7Q2	3.58	Peptide;H2O	mutant SPase I  R127A	15195997	Escherichia coli	P00803	3.4.21.89	kcat/Km	kcat/Km	Peptide	37.63889		-	M^(-1)*s^(-1)	7.7	37.0

EntryID: 65338
65338	Oxaloacetate;H2O;Acetyl-CoA	wildtype	10091606	Helicobacter pylori	P56062	2.3.3.16	Km_B	Km	Oxaloacetate	3.47E-4		4.3E-5	M	7.4	37.654	UDPglucuronate;Cyanidin 3-O-(6-O-malonyl-beta-D-glucoside)	wildtype	15509561	Bellis perennis	Q5NTH0	2.4.1.17	KmB	Km	UDPglucuronate	4.97E-4		9.1E-5	M	7.0	30.0

EntryID: 68232
68232	Arg-Leu-Arg-Gly-Gly-AMC;H2O	wildtype PLpro (aa 1541-1855)	19645480	SARS coronavirus	P0C6U8 P0C6X7	3.4.22.n14	I	concentration	5-Amino-2-methyl-N-[(1R)-1-(1-naphthyl)ethyl]benzamide	0.0	2.0E-4	-	M	7.5	683	6-Phospho-D-gluconate;NADP+	wildtype	3965621	Homo sapiens		1.1.1.44	B	concentration	6-Phospho-D-gluconate	3.0E-5	1.5E-4	-	M	7.0	-

EntryID: 71116
71116	NADH;ATP	wildtype	15855156	Micrococcus luteus	Q8VUL9	2.7.1.86	A	concentration	NADH	0712	

In [18]:
QUERY_URL = 'http://sabiork.h-its.org/sabioRestWebServices/kineticlawsExportTsv'
query = {'fields[]':['EntryID', 'Substrate', 'EnzymeType', 'PubMedID', 'Organism', 'UniprotID', 'ECNumber', 'Parameter', 'pH', 'Temperature']}

In [22]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:57731"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

2.330843925476074
EntryID	Substrate	EnzymeType	PubMedID	Organism	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
57731	H2O;GM3	wildtype	7325957	Paenarthrobacter ureafaciens	Q5W7Q2	3.2.1.18	Vmax	Vmax		1.95E-4		-	mol*s^(-1)*g^(-1)	4.8	37.0
57731	H2O;GM3	wildtype	7325957	Paenarthrobacter ureafaciens	Q5W7Q2	3.2.1.18	Km	Km	GM3	8.0E-5		-	M	4.8	37.0



In [23]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:65338"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

7.627881050109863
EntryID	Substrate	EnzymeType	PubMedID	Organism	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
65338	H2O;Acetyl-CoA;Oxaloacetate	wildtype	10091606	Helicobacter pylori	P56062	2.3.3.16	B	concentration	Oxaloacetate	0.005		-	M	7.4	37.0
65338	H2O;Acetyl-CoA;Oxaloacetate	wildtype	10091606	Helicobacter pylori	P56062	2.3.3.16	A	concentration	Acetyl-CoA	0.005		-	M	7.4	37.0
65338	H2O;Acetyl-CoA;Oxaloacetate	wildtype	10091606	Helicobacter pylori	P56062	2.3.3.16	Km_B	Km	Oxaloacetate	3.47E-4		4.3E-5	M	7.4	37.0
65338	H2O;Acetyl-CoA;Oxaloacetate	wildtype	10091606	Helicobacter pylori	P56062	2.3.3.16	Vmax	Vmax		2.816667E-6		1.833333E-7	mol*s^(-1)*g^(-1)	7.4	37.0
65338	H2O;Acetyl-CoA;Oxaloacetate	wildtype	10091606	Helicobacter pylori	P56062	2.3.3.16	Km_A	Km	Acetyl-CoA	2.04E-4		3.4E-5	M	7.4	37.0
65338	H2O;Acetyl-CoA;Oxaloacetate	wildtype	10091606	Helicobacter py

In [24]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:68232"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

5.917498826980591
EntryID	Substrate	EnzymeType	PubMedID	Organism	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
68232	H2O;Arg-Leu-Arg-Gly-Gly-AMC	wildtype PLpro (aa 1541-1855)	19645480	SARS coronavirus	P0C6U8 P0C6X7	3.4.22.n14	IC50	IC50	5-Amino-2-methyl-N-[(1R)-1-(1-naphthyl)ethyl]benzamide	5.6E-7		3E-8	M	7.5	-
68232	H2O;Arg-Leu-Arg-Gly-Gly-AMC	wildtype PLpro (aa 1541-1855)	19645480	SARS coronavirus	P0C6U8 P0C6X7	3.4.22.n14	E	concentration	unknown compound	1.25E-7		-	M	7.5	-
68232	H2O;Arg-Leu-Arg-Gly-Gly-AMC	wildtype PLpro (aa 1541-1855)	19645480	SARS coronavirus	P0C6U8 P0C6X7	3.4.22.n14	vi	rate const.				-	-	7.5	-
68232	H2O;Arg-Leu-Arg-Gly-Gly-AMC	wildtype PLpro (aa 1541-1855)	19645480	SARS coronavirus	P0C6U8 P0C6X7	3.4.22.n14	I	concentration	5-Amino-2-methyl-N-[(1R)-1-(1-naphthyl)ethyl]benzamide	0.0	2.0E-4	-	M	7.5	-
68232	H2O;Arg-Leu-Arg-Gly-Gly-AMC	wildtype

In [25]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:71116"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

6.0103020668029785
EntryID	Substrate	EnzymeType	PubMedID	Organism	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
71116	NADH;ATP	wildtype	15855156	Micrococcus luteus	Q8VUL9	2.7.1.86	Km	Km	ATP	3.2E-4		6E-6	M	7.0	37.0
71116	NADH;ATP	wildtype	15855156	Micrococcus luteus	Q8VUL9	2.7.1.86	E	concentration	unknown compound	0.004		-	mg/ml	7.0	37.0
71116	NADH;ATP	wildtype	15855156	Micrococcus luteus	Q8VUL9	2.7.1.86	S	concentration	ATP			-	-	7.0	37.0
71116	NADH;ATP	wildtype	15855156	Micrococcus luteus	Q8VUL9	2.7.1.86	Vmax	Vmax		1.441667E-4		4.85E-6	mol*s^(-1)*g^(-1)	7.0	37.0
71116	NADH;ATP	wildtype	15855156	Micrococcus luteus	Q8VUL9	2.7.1.86	A	concentration	NADH	0.0025		-	M	7.0	37.0



In [26]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:75485"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

7.0985589027404785
EntryID	Substrate	EnzymeType	PubMedID	Organism	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
75485	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	29807982	Papaver somniferum	A0A2U9GGW3	4.2.99.24	S	concentration	7-O-Acetylsalutaridinol	0.0	1.0E-4	-	M	7.0	30.0
75485	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	29807982	Papaver somniferum	A0A2U9GGW3	4.2.99.24	Vmax	Vmax		5.36E-5		3.05E-6	mol*s^(-1)*g^(-1)	7.0	30.0
75485	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	29807982	Papaver somniferum	A0A2U9GGW3	4.2.99.24	n	Hill coefficient	7-O-Acetylsalutaridinol	2.316		3.52E-1	-	7.0	30.0
75485	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	29807982	Papaver somniferum	A0A2U9GGW3	4.2.99.24	S_half	S_half	7-O-Acetylsalutaridinol	2.197E-5		2.036E-6	M	7.0	30.0



In [27]:
for line in lines:
    data = line.strip().split('\t')
    if len(data) != 16:
        print('EntryID:', data[0])
        print(line)

EntryID: 57731
57731	H2O;GM3	wildtype	7325957	Paenarthrobacter ureafaciens	Q5W7Q2	3.58	Peptide;H2O	mutant SPase I  R127A	15195997	Escherichia coli	P00803	3.4.21.89	kcat/Km	kcat/Km	Peptide	37.63889		-	M^(-1)*s^(-1)	7.7	37.0

EntryID: 65338
65338	Oxaloacetate;H2O;Acetyl-CoA	wildtype	10091606	Helicobacter pylori	P56062	2.3.3.16	Km_B	Km	Oxaloacetate	3.47E-4		4.3E-5	M	7.4	37.654	UDPglucuronate;Cyanidin 3-O-(6-O-malonyl-beta-D-glucoside)	wildtype	15509561	Bellis perennis	Q5NTH0	2.4.1.17	KmB	Km	UDPglucuronate	4.97E-4		9.1E-5	M	7.0	30.0

EntryID: 68232
68232	Arg-Leu-Arg-Gly-Gly-AMC;H2O	wildtype PLpro (aa 1541-1855)	19645480	SARS coronavirus	P0C6U8 P0C6X7	3.4.22.n14	I	concentration	5-Amino-2-methyl-N-[(1R)-1-(1-naphthyl)ethyl]benzamide	0.0	2.0E-4	-	M	7.5	683	6-Phospho-D-gluconate;NADP+	wildtype	3965621	Homo sapiens		1.1.1.44	B	concentration	6-Phospho-D-gluconate	3.0E-5	1.5E-4	-	M	7.0	-

EntryID: 71116
71116	NADH;ATP	wildtype	15855156	Micrococcus luteus	Q8VUL9	2.7.1.86	A	concentration	NADH	0712	

### Resolve inconsistent formatting

In [ ]:
fdir = '../../../../data/databases/Sabio-RK/'
fname = 'dataset_download.tsv'
with open(fdir+fname, 'r') as f:
    data = f.read()

In [2]:
# remove unnecessary dashes and spaces so the corresponding entries are treated as missing values
data = data.replace("\t-", "\t")
data = data.replace("\t ", "\t")
data = data.replace("\t-\n", "\t\n")

In [3]:
fname = 'Sabio-RK_dataset.tsv'
with open(fdir+fname, 'w') as f:
    f.write(data)

# Extract and clean up $k_{cat}$ entries

### Load data

Load the dataset as pandas dataframe and inspect the specifics of each parameter and their associated values

In [4]:
import pandas as pd
import numpy as np
pd.options.mode.chained_assignment = None

In [5]:
header = ['EntryID', 'Substrate', 'EnzymeType', 'PubMedID', 'Organism', 'UniprotID', 'ECNumber',
           'parameter.name', 'parameter.type', 'parameter.associatedSpecies', 'parameter.startValue', 'parameter.endValue', 'parameter.standardDeviation', 'parameter.unit',
           'pH', 'Temperature']

In [ ]:
fdir = '../../../../data/databases/Sabio-RK/'
fname = 'Sabio-RK_dataset.tsv'
df = pd.read_csv(fdir+fname, sep='\t', names=header)

In [7]:
len(np.unique(df['EntryID']))

74395

In [8]:
df

EntryID                             Substrate  \
0             1                           Peptide;H2O   
1             1                           Peptide;H2O   
2             1                           Peptide;H2O   
3             1                           Peptide;H2O   
4            10  (S)-Mandelate;Riboflavin-5-phosphate   
...         ...                                   ...   
334081     9997                  2-Deoxy-D-glucose;O2   
334082     9998      O2;beta-Methyl-D-glucopyranoside   
334083     9998      O2;beta-Methyl-D-glucopyranoside   
334084     9998      O2;beta-Methyl-D-glucopyranoside   
334085     9999      beta-Methyl-D-glucopyranoside;O2   

                                           EnzymeType    PubMedID  \
0       wildtype variant DSAI (N76D/N87S/S103A/V104I)  12962477.0   
1       wildtype variant DSAI (N76D/N87S/S103A/V104I)  12962477.0   
2       wildtype variant DSAI (N76D/N87S/S103A/V104I)  12962477.0   
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)  12962477.0   
4                                     mutant MDH G81S  15311930.0   
...                                               ...         ...   
334081              mutant P2OxB2H (K312E-E540K-His6)  15660220.0   
334082                                  wildtype P2Ox  15660220.0   
334083                                  wildtype P2Ox  15660220.0   
334084                                  wildtype P2Ox  15660220.0   
334085                          mutant P2OxB1 (E540K)  15660220.0   

                    Organism UniprotID   ECNumber parameter.name  \
0            Bacillus lentus       NaN  3.4.21.62             Km   
1            Bacillus lentus       NaN  3.4.21.62             Ki   
2            Bacillus lentus       NaN  3.4.21.62        kcat/Km   
3            Bacillus lentus       NaN  3.4.21.62           kcat   
4         Pseudomonas putida    P20932  1.1.99.31             Km   
...                      ...       ...        ...            ...   
334081  Phlebiopsis gigantea    Q6UG02   1.1.3.10             Km   
334082  Phlebiopsis gigantea    Q6UG02   1.1.3.10        kcat_Km   
334083  Phlebiopsis gigantea    Q6UG02   1.1.3.10           Vmax   
334084  Phlebiopsis gigantea    Q6UG02   1.1.3.10             Km   
334085  Phlebiopsis gigantea    Q6UG02   1.1.3.10             Km   

       parameter.type                        parameter.associatedSpecies  \
0                  Km                                            Peptide   
1                  Ki  N-(Carbobenzoxy)-Leu-Leu-Phe-trifluoromethylke...   
2             kcat/Km                                            Peptide   
3                kcat                                                NaN   
4                  Km                                      (S)-Mandelate   
...               ...                                                ...   
334081             Km                                  2-Deoxy-D-glucose   
334082        kcat/Km                      beta-Methyl-D-glucopyranoside   
334083           Vmax                                                NaN   
334084             Km                      beta-Methyl-D-glucopyranoside   
334085             Km                      beta-Methyl-D-glucopyranoside   

        parameter.startValue  parameter.endValue  parameter.standardDeviation  \
0               1.900000e-03                 NaN                          NaN   
1               1.800000e-08                 NaN                          NaN   
2               1.200000e+05                 NaN                          NaN   
3               2.200000e+02                 NaN                          NaN   
4               2.300000e-03                 NaN                       0.0002   
...                      ...                 ...                          ...   
334081          9.930000e-02                 NaN                       0.0040   
334082          1.680000e+02                 NaN                          NaN   
334083          4.833333e-05       

### Extract $k_{cat}$ values

In [9]:
mask = df['parameter.type'] == 'kcat'
df = df.loc[mask]
n0 = len(df)
n0

29588

Remove missing $k_{cat}$ values

In [10]:
mask = ~df['parameter.startValue'].isnull()
print("Number of entries with no assigned value:", n0 - np.sum(mask))
df = df.loc[mask]

Number of entries with no assigned value: 1368


In [11]:
print("Remaining number of entries:", len(df))

Remaining number of entries: 28220


Remove unnecessary `parameter.name`, `parameter.type` and `parameter.associatedSpecies` columns

In [12]:
df = df.drop(['parameter.name', 'parameter.type', 'parameter.associatedSpecies'], axis=1)

In [13]:
df

EntryID                                  Substrate  \
3             1                                Peptide;H2O   
6            10       (S)-Mandelate;Riboflavin-5-phosphate   
8           100                     NADPH;H+;Dihydrofolate   
13          101                     H+;NADPH;Dihydrofolate   
18          102                     NADPH;Dihydrofolate;H+   
...         ...                                        ...   
333509     9874                 H2O;O2;N,N-Dimethylglycine   
333511     9875                 O2;H2O;N,N-Dimethylglycine   
333513     9876                 O2;N,N-Dimethylglycine;H2O   
333972     9970  5alpha-Androstane-17beta-ol-3-one;NADH;H+   
333980     9971  NAD+;5alpha-Androstane-3alpha,17beta-diol   

                                           EnzymeType    PubMedID  \
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)  12962477.0   
6                                     mutant MDH G81S  15311930.0   
8                                       wildtype DHFR  15740745.0   
13                                   mutant DHFR W30A  15740745.0   
18                                   mutant DHFR W30E  15740745.0   
...                                               ...         ...   
333509                                       wildtype  12912903.0   
333511                                   mutant H225Q  12912903.0   
333513                                   mutant Y259F  12912903.0   
333972                                  wildtype P26h  11141073.0   
333980                                  wildtype P26h  11141073.0   

                        Organism UniprotID   ECNumber  parameter.startValue  \
3                Bacillus lentus       NaN  3.4.21.62              220.0000   
6             Pseudomonas putida    P20932  1.1.99.31                2.8000   
8               Escherichia coli    P0ABQ4    1.5.1.3                6.1800   
13              Escherichia coli    P0ABQ4    1.5.1.3               14.5000   
18              Escherichia coli    P0ABQ4    1.5.1.3               11.5800   
...                          ...       ...        ...                   ...   
333509  Arthrobacter globiformis       NaN   1.5.3.10               10.6000   
333511  Arthrobacter globiformis       NaN   1.5.3.10                4.3500   
333513  Arthrobacter globiformis       NaN   1.5.3.10                0.0047   
333972      Mesocricetus auratus    Q9Z245  1.1.1.184                0.8500   
333980      Mesocricetus auratus    Q9Z245  1.1.1.184                1.2800   

        parameter.endValue  parameter.standardDeviation parameter.unit   pH  \
3                      NaN                          NaN         s^(-1)  7.5   
6                      NaN                          0.1         s^(-1)  7.5   
8                      NaN                          NaN         s^(-1)  7.0   
13                     NaN                          NaN         s^(-1)  7.0   
18                     NaN                          NaN         s^(-1)  7.0   
...                    ...                          ...            ...  ...   
333509                 NaN                          NaN         s^(-1)  8.5   
333511                 NaN                          NaN         s^(-1)  8.5   
333513                 NaN                          NaN         s^(-1)  8.5   
333972                 NaN                          NaN         s^(-1)  7.0   
333980                 NaN                          NaN         s^(-1)  7.0   

        Temperature  
3              25.0  
6              20.0  
8              15.0  
13             15.0  
18             15.0  
...             ...  
333509         25.0  
333511         25.0  
333513         25.0  
333972         25.0  
333980         25.0  

[28220 rows x 13 columns]

### Explore and reformat data entries

#### `EntryID`

In [14]:
len(np.unique(df['EntryID']))

28056

Only 28k out of all 74k EntryIDs in the database have an associated $k_{cat}$ value.

#### `PubMedID`

These are parsed as floats but should be integers

In [15]:
np.unique(df['PubMedID'])

array([8.4470000e+03, 1.6919000e+04, 1.9039000e+04, ..., 3.3810098e+07,
       3.5092420e+07,           nan])

In [16]:
df['PubMedID']

3         12962477.0
6         15311930.0
8         15740745.0
13        15740745.0
18        15740745.0
             ...    
333509    12912903.0
333511    12912903.0
333513    12912903.0
333972    11141073.0
333980    11141073.0
Name: PubMedID, Length: 28220, dtype: float64

In [17]:
df['PubMedID'] = df['PubMedID'].astype('Int64')

#### `Substrate`

In the end we reduce the dataset to $k_{cat}$ values only. If I understand correctly, in the DLKcat paper, the authors associate the $k_{cat}$ value with the substrate that has a measured $K_M$ value. If the $K_M$ values are measured for multiple substrates in the same reaction, then this would result in separate data inputs of the form $\text{substrate}_1 \rightarrow k_{cat}$ and $\text{substrate}_2 \rightarrow k_{cat}$ (in the neural net framework this can lead to different `kcat` predictions depending on the input substrate). We want to construct an input vector that includes all the substrates involved in the reaction (ignoring whether it has a measured $K_M$ value) in the form $[\text{substrate}_1, \text{substrate}_2, \dotsc] \rightarrow k_{cat}$. 

This is mentioned in the paper's discussion: "Another challenge relates to reactions involving multiple substrates and those catalysed by heteromeric enzyme complexes. The multiple substrate SMILES and protein sequences that can be defined for such reactions can all function with DLKcat, thereby yielding multiple predicted kcat values for one reaction. We currently select the maximum kcat values in those cases, but it would be favourable to devise an approach that can predict one kcat value for each multi-substrate and/or heteromeric enzyme."

Each entry is a list of substrates separated by `;` -- split into an array of individual substrate strings.

In [18]:
df['Substrate'] = df['Substrate'].apply(lambda x: np.sort(x.split(';')))

#### `EnzymeType`

In [19]:
df['EnzymeType']

3         wildtype variant DSAI (N76D/N87S/S103A/V104I)
6                                       mutant MDH G81S
8                                         wildtype DHFR
13                                     mutant DHFR W30A
18                                     mutant DHFR W30E
                              ...                      
333509                                         wildtype
333511                                     mutant H225Q
333513                                     mutant Y259F
333972                                    wildtype P26h
333980                                    wildtype P26h
Name: EnzymeType, Length: 28220, dtype: object

In [20]:
mutant_namelist = df['EnzymeType'][df['EnzymeType'].astype(str).apply(lambda x: 'mutant' in x)]

In [21]:
# code from DLKcat
import re

for name in df['EnzymeType'].astype(str): 
    if 'wildtype' in name:
        print (name +' -> wildtype')
    else :
    # if 'mutant' in EnzymeType or 'mutated' in EnzymeType:
        mutant = re.findall('[A-Z]\d+[A-Z]', name)  # re is of great use
        print(name + ' -> '+ '/'.join(mutant))

wildtype variant DSAI (N76D/N87S/S103A/V104I) -> wildtype
mutant MDH G81S -> G81S
wildtype DHFR -> wildtype
mutant DHFR W30A -> W30A
mutant DHFR W30E -> W30E
mutant DHFR W30H -> W30H
wildtype -> wildtype
wildtype -> wildtype
wildtype -> wildtype
wildtype -> wildtype
mutant Y538F -> Y538F
mutant E637D -> E637D
wildtype -> wildtype
mutant Y538F -> Y538F
mutant E637D -> E637D
wildtype -> wildtype
mutant Y538F -> Y538F
mutant E637D -> E637D
wildtype -> wildtype
mutant Y538F -> Y538F
mutant E637D -> E637D
wildtype -> wildtype
mutant Y538F -> Y538F
mutant E637D -> E637D
wildtype -> wildtype
mutant Y538F -> Y538F
mutant E637D -> E637D
wildtype -> wildtype
wildtype -> wildtype
wildtype -> wildtype
wildtype -> wildtype
wildtype MenD -> wildtype
wildtype MenD -> wildtype
wildtype MenD -> wildtype
wildtype MenD -> wildtype
wildtype -> wildtype
mutant S73A -> S73A
mutant Y74F -> Y74F
mutant Y74E -> Y74E
mutant Y74W -> Y74W
mutant S75H -> S75H
wildtype -> wildtype
mutant S96A -> S96A
mutant Y97F ->

<>:9: SyntaxWarning: invalid escape sequence '\d'
<>:9: SyntaxWarning: invalid escape sequence '\d'
/var/folders/mm/scs9kbl94v502ndl7bjxwsp00000gn/T/ipykernel_92848/56168841.py:9: SyntaxWarning: invalid escape sequence '\d'
  mutant = re.findall('[A-Z]\d+[A-Z]', name)  # re is of great use


The modifications given above that simplify the mutant/wild-type descriptors are performed in the DLKcat scripts. Omitting it now to keep things general.

#### `Organism`

In [22]:
np.unique(df['Organism'])

array(['Abax parallelepipedus', 'Acanthamoeba polyphaga mimivirus',
       'Acetobacter aceti', 'Acetobacter xylinus',
       'Achromobacter cycloclastes', 'Achromobacter xylosoxidans',
       'Acidianus sp.', 'Acidithiobacillus thiooxidans',
       'Acidobacteriaceae', 'Acidovorax citrulli',
       'Acinetobacter baumannii', 'Acinetobacter baylyi',
       'Acinetobacter calcoaceticus',
       'Acinetobacter calcoaceticus (subsp. anitratus)',
       'Acremonium persicinum', 'Actinidia chinensis',
       'Actinidia eriantha', 'Actinobacillus succinogenes',
       'Actinomadura sp.', 'Actinoplanes missouriensis', 'Aedes aegypti',
       'Aeromonas hydrophila', 'Aeromonas sobria', 'Aeropyrum pernix',
       'Agaricus bisporus', 'Agarivorans albus', 'Agkistrodon bilineatus',
       'Agkistrodon rhodostoma', 'Agrobacterium tumefaciens',
       'Agrobacterium vitis', 'Agrocybe aegerita',
       'Akkermansia muciniphila', 'Alcaligenes faecalis',
       'Alicyclobacillus acidocaldarius',
     

Keeping the organism list as is.

#### `UniprotID`

In [23]:
np.unique(df['UniprotID'][df['UniprotID'].apply(lambda x: type(x) != str)].astype(str))

array(['nan'], dtype=object)

Some entries do not have an associated UniprotID.

In [24]:
df.loc[[3]]

EntryID       Substrate                                     EnzymeType  \
3        1  [H2O, Peptide]  wildtype variant DSAI (N76D/N87S/S103A/V104I)   

   PubMedID         Organism UniprotID   ECNumber  parameter.startValue  \
3  12962477  Bacillus lentus       NaN  3.4.21.62                 220.0   

   parameter.endValue  parameter.standardDeviation parameter.unit   pH  \
3                 NaN                          NaN         s^(-1)  7.5   

   Temperature  
3         25.0

In [25]:
df.loc[[313665]]

EntryID     Substrate                   EnzymeType  PubMedID  \
313665    75307  [L-Tyrosine]  wildtype mfnA_2 His6-tagged  30659181   

                    Organism UniprotID  ECNumber  parameter.startValue  \
313665  Enterococcus faecium       NaN  4.1.1.25              36.73333   

        parameter.endValue  parameter.standardDeviation parameter.unit   pH  \
313665                 NaN                     4.116667         s^(-1)  5.0   

        Temperature  
313665          NaN

Some entries have multiple associated UniprotIDs (represent heteromeric protein complexes?). Detect such entries and list the UniprotIDs as an array

In [26]:
df['UniprotID'] = df['UniprotID'].apply(lambda x: np.empty(0) if pd.isnull(x) else np.sort(x.split(' ')))

In [27]:
df[df['UniprotID'].astype(str).apply(lambda x : len(x)) == 2]

EntryID                                          Substrate  \
3             1                                     [H2O, Peptide]   
391       10048    [H2O, trans-4-Carboxymethylenebut-2-en-4-olide]   
397       10049                       [4-Fluoromuconolactone, H2O]   
529       10082  [5-Phospho-alpha-D-ribose 1-diphosphate, Guanine]   
532       10083  [5-Phospho-alpha-D-ribose 1-diphosphate, Guanine]   
...         ...                                                ...   
332147      990            [Dehydroascorbate, Reduced glutathione]   
333083     9793            [GDP-4-dehydro-D-rhamnose, L-Glutamate]   
333509     9874                     [H2O, N,N-Dimethylglycine, O2]   
333511     9875                     [H2O, N,N-Dimethylglycine, O2]   
333513     9876                     [H2O, N,N-Dimethylglycine, O2]   

                                           EnzymeType  PubMedID  \
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)  12962477   
391                                          wildtype  14617643   
397                                          wildtype  14617643   
529                                          wildtype  11284686   
532                                       mutant S96A  11284686   
...                                               ...       ...   
332147                                       wildtype   9890907   
333083                                  wildtype ColD  15610039   
333509                                       wildtype  12912903   
333511                                   mutant H225Q  12912903   
333513                                   mutant Y259F  12912903   

                           Organism UniprotID   ECNumber  \
3                   Bacillus lentus        []  3.4.21.62   
391            Pseudomonas reinekei        []   3.1.1.45   
397            Pseudomonas reinekei        []   3.1.1.45   
529                 Giardia lamblia        []      2.4.2   
532                 Giardia lamblia        []      2.4.2   
...                             ...       ...        ...   
332147                   Sus scrofa        []    1.8.5.1   
333083  Yersinia pseudotuberculosis        []        NaN   
333509     Arthrobacter globiformis        []   1.5.3.10   
333511     Arthrobacter globiformis        []   1.5.3.10   
333513     Arthrobacter globiformis        []   1.5.3.10   

        parameter.startValue  parameter.endValue  parameter.standardDeviation  \
3                 220.000000                 NaN                          NaN   
391              2650.000000                 NaN                   141.666700   
397              1633.333000                 NaN                    36.666670   
529                76.700000                 NaN                     2.500000   
532                55.700000                 NaN                     4.200000   
...                      ...                 ...                          ...   
332147              7.566667                 NaN                     0.283333   
333083              0.600000                 NaN                          NaN   
333509             10.600000                 NaN                          NaN   
333511              4.350000                 NaN                          NaN   
333513              0.004700                 NaN                          NaN   

       parameter.unit   pH  Temperature  
3              s^(-1)  7.5         25.0  
391            s^(-1)  7.5         25.0  
397            s^(-1)  7.5         25.0  
529            s^(-1)  7.0         37.0  
532            s^(-1)  7.0         37.0  
...               ...  ...          ...  
332147         s^(-1)  6.9         30.0  
333083         s^(-1)  7.0         37.0  
333509         s^(-1)  8.5         25.0  
333511         s^(-1)  8.5         25.0  
333513         s^(-1)  8.5         25.0  

[2151 rows x 13 columns]

#### `ECNumber`

Multiple ECNumbers may be associated with an entry -- putting them in arrays.

In [28]:
df['ECNumber'] = df['ECNumber'].apply(lambda x: np.empty(0) if pd.isnull(x) else np.sort(x.split(' ')))

In [29]:
mask = (df['ECNumber'].apply(lambda x: len(x) == 0)) & (df['UniprotID'].apply(lambda x: len(x) == 0))

In [30]:
np.sum(mask)

20

Remove entries that are missing both `ECNumber` and `UniprotID` (have no associated enzyme information) 

In [31]:
df = df.loc[~mask]

#### `parameter`

Some database entries have multiple associated $k_{cat}$ values. In DLKcat, the duplicates are removed by selecting the maximum $k_{cat}$ value as the ground truth. This is not necessarily the best approach as the multiple values may reflect experimental uncertainty or alternative specific experimental conditions. However, this is difficult to parse in an automated way, and dealing with such values more rigorously may require manual labour...

Finally, note that in some cases the $k_{cat}$ values have not only an associated `parameter.startValue` but also an extra `parameter.endValue` (indicating an interval of possible values) or an associated `parameter.standardDeviation`. It's questionable whether these could be used as extra inputs.

In [32]:
i = 0
for entryID in np.unique(df['EntryID']):
    n = np.sum(df['EntryID'] == entryID)
    if n > 1:
        i += 1
        print('EntryID %d has %d kcat values' % (entryID, n))
print('Number of entries with more than one kcat: %d' %i)

EntryID 294 has 2 kcat values
EntryID 295 has 2 kcat values
EntryID 296 has 2 kcat values
EntryID 297 has 2 kcat values
EntryID 428 has 3 kcat values
EntryID 443 has 2 kcat values
EntryID 444 has 2 kcat values
EntryID 445 has 2 kcat values
EntryID 486 has 2 kcat values
EntryID 985 has 2 kcat values
EntryID 986 has 2 kcat values
EntryID 1075 has 2 kcat values
EntryID 1868 has 2 kcat values
EntryID 1869 has 2 kcat values
EntryID 1870 has 2 kcat values
EntryID 1871 has 2 kcat values
EntryID 1973 has 2 kcat values
EntryID 1974 has 2 kcat values
EntryID 1975 has 2 kcat values
EntryID 1976 has 2 kcat values
EntryID 1977 has 2 kcat values
EntryID 1978 has 2 kcat values
EntryID 1979 has 2 kcat values
EntryID 1980 has 2 kcat values
EntryID 1981 has 2 kcat values
EntryID 1982 has 2 kcat values
EntryID 1983 has 2 kcat values
EntryID 1984 has 2 kcat values
EntryID 2872 has 2 kcat values
EntryID 2873 has 2 kcat values
EntryID 2874 has 2 kcat values
EntryID 2875 has 2 kcat values
EntryID 2876 has 2 

A few are due to duplicated data entries. Remove those

In [33]:
df = df.loc[df.astype(str).drop_duplicates().index]

In [184]:
len(df)

28195

In [34]:
i = 0
for entryID in np.unique(df['EntryID']):
    n = np.sum(df['EntryID'] == entryID)
    if n > 1:
        i += 1
        print('EntryID %d has %d kcat values' % (entryID, n))
print('Number of entries with more than one kcat: %d' %i)

EntryID 294 has 2 kcat values
EntryID 295 has 2 kcat values
EntryID 296 has 2 kcat values
EntryID 297 has 2 kcat values
EntryID 428 has 3 kcat values
EntryID 443 has 2 kcat values
EntryID 444 has 2 kcat values
EntryID 445 has 2 kcat values
EntryID 486 has 2 kcat values
EntryID 985 has 2 kcat values
EntryID 986 has 2 kcat values
EntryID 1075 has 2 kcat values
EntryID 1868 has 2 kcat values
EntryID 1869 has 2 kcat values
EntryID 1870 has 2 kcat values
EntryID 1871 has 2 kcat values
EntryID 1973 has 2 kcat values
EntryID 1974 has 2 kcat values
EntryID 1975 has 2 kcat values
EntryID 1976 has 2 kcat values
EntryID 1977 has 2 kcat values
EntryID 1978 has 2 kcat values
EntryID 1979 has 2 kcat values
EntryID 1980 has 2 kcat values
EntryID 1981 has 2 kcat values
EntryID 1982 has 2 kcat values
EntryID 1983 has 2 kcat values
EntryID 1984 has 2 kcat values
EntryID 2873 has 2 kcat values
EntryID 2874 has 2 kcat values
EntryID 2875 has 2 kcat values
EntryID 2876 has 2 kcat values
EntryID 2878 has 2 

#### `parameter.endValue`

List the number of $k_{cat}$ values with an associated `endValue`

In [35]:
np.sum(~df['parameter.endValue'].isnull())

103

Very few entries exist, so ignoring `endValue` seems to be the way to go. Remove the column completely

In [36]:
df = df.drop('parameter.endValue', axis=1)

#### `parameter.standardDeviation`

List the number of $k_{cat}$ values with an associated `standardDeviation`

In [37]:
np.sum(~df['parameter.standardDeviation'].isnull())

16020

Standard deviation is given for a relatively good chunk -- might be worth considering.

#### `parameter.unit`

Units are not given in certain cases.

In [38]:
df[df['parameter.unit'].apply(lambda x : pd.isnull(x))]

EntryID                                Substrate  \
149830    41967  [3'-Phosphoadenylyl sulfate, Quercetin]   
149839    41969  [3'-Phosphoadenylyl sulfate, Quercetin]   
149852    41971  [3'-Phosphoadenylyl sulfate, Quercetin]   
149863    41973  [3'-Phosphoadenylyl sulfate, Quercetin]   
149877    41975  [3'-Phosphoadenylyl sulfate, Quercetin]   
149890    41977  [3'-Phosphoadenylyl sulfate, Quercetin]   
149898    41979  [3'-Phosphoadenylyl sulfate, Quercetin]   
149912    41981  [3'-Phosphoadenylyl sulfate, Quercetin]   
149924    41983  [3'-Phosphoadenylyl sulfate, Quercetin]   
149934    41985  [3'-Phosphoadenylyl sulfate, Quercetin]   
204797    53222                      [H2O, Penicillin V]   
303616    73389       [Dabcyl-KTSAVLQSGFRKME-Edans, H2O]   

                            EnzymeType  PubMedID                Organism  \
149830                     mutant P120   9288931   Flaveria chloraefolia   
149839                     mutant L120   9288931   Flaveria chloraefolia   
149852                     mutant K126   9288931   Flaveria chloraefolia   
149863                     mutant R134   9288931   Flaveria chloraefolia   
149877                     mutant T134   9288931   Flaveria chloraefolia   
149890                     mutant G134   9288931   Flaveria chloraefolia   
149898                     mutant A120   9288931   Flaveria chloraefolia   
149912                     mutant H175   9288931   Flaveria chloraefolia   
149924                     mutant Q186   9288931   Flaveria chloraefolia   
149934                     mutant A186   9288931   Flaveria chloraefolia   
204797                    mutant K234E   1974463  Bacillus licheniformis   
303616  wildtype 3CLpro (aa 3241-3546)  18237196        SARS coronavirus   

                               UniprotID     ECNumber  parameter.startValue  \
149830                          [P52836]   [2.8.2.25]                  0.53   
149839                          [P52836]   [2.8.2.25]                  0.69   
149852                          [P52836]   [2.8.2.25]                  0.92   
149863                          [P52836]   [2.8.2.25]                  1.66   
149877                          [P52836]   [2.8.2.25]                  0.53   
149890                          [P52836]   [2.8.2.25]                  0.27   
149898                          [P52836]   [2.8.2.25]                  0.41   
149912                          [P52836]   [2.8.2.25]                  1.36   
149924                          [P52836]   [2.8.2.25]                  0.26   
149934                          [P52836]   [2.8.2.25]                  0.06   
204797                          [P00808]    [3.5.2.6]                 37.30   
303616  [P0C6U8, P0C6U8, P0C6X7, P0C6X7]  [3.4.22.69]                  3.70   

        parameter.standardDeviation parameter.unit   pH  Temperature  
149830                          NaN            NaN  7.5         25.0  
149839                          NaN            NaN  7.5         25.0  
149852                          NaN            NaN  7.5         25.0  
149863                          NaN            NaN  7.5         25.0  
149877                          NaN            NaN  7.5         25.0  
149890                          NaN            NaN  7.5         25.0  
149898                          NaN            NaN  7.5         25.0  
149912                          NaN            NaN  7.5         25.0  
149924                          NaN            NaN  7.5         25.0  
149934                          NaN            NaN  7.5         25.0  
204797                          NaN            NaN  7.0         30.0  
303616                          0.1            NaN  7.0         22.0

Having checked the paper for EntryID: 41969, the units are given as $s^{-1}$, so it's a human error in this case. Probably okay to assume that all `NaN` units can be converted to $s^{-1}$ (as done in DLKcat). 

Some $k_{cat}$ values have weird units associated with them. Some appear to be an error, also leading to wrong values due to an automatic unit conversion step.

In [39]:
np.unique(df['parameter.unit'].astype(str))

array(['J/mol', 'M', 'M^(-1)*s^(-1)', 'katal', 'mol*s^(-1)*g^(-1)',
       'mol*s^(-1)*mol^(-1)', 'nan', 's^(-1)', 's^(-1)*g^(-1)'],
      dtype=object)

In [40]:
df[df['parameter.unit'] == 'J/mol']

EntryID         Substrate           EnzymeType  PubMedID  \
270390    66706  [ATP, Thymidine]  wildtype His-tagged  17335913   

              Organism UniprotID    ECNumber  parameter.startValue  \
270390  Vaccinia virus  [P68563]  [2.7.1.21]              0.921096   

        parameter.standardDeviation parameter.unit   pH  Temperature  
270390                     0.167472          J/mol  8.0         37.0

Typo having checked the paper (should be $s^{-1}$)

In [41]:
df[df['parameter.unit'] == 'M']

EntryID                                    Substrate  \
110272    33792     [5alpha-Androstane-3,17-dione, H+, NADH]   
110291    33796     [5alpha-Androstane-3,17-dione, H+, NADH]   
110310    33800     [5alpha-Androstane-3,17-dione, H+, NADH]   
110332    33804     [5alpha-Androstane-3,17-dione, H+, NADH]   
200400    52323        [Glp-Ala-Ala-Leu-p-Nitroanilide, H2O]   
200406    52324        [Glp-Ala-Ala-Leu-p-Nitroanilide, H2O]   
254553    63591  [Enzyme-cysadenylate-PPi complex, GlcN-Ins]   

                                 EnzymeType  PubMedID  \
110272    mutant isoenzyme 3alpha-HSD C170A   8175784   
110291    mutant isoenzyme 3alpha-HSD C217A   8175784   
110310    mutant isoenzyme 3alpha-HSD C242A   8175784   
110332    mutant isoenzyme 3alpha-HSD Y205F   8175784   
200400                             wildtype  14640970   
200406  wildtype poly(acrylic acid) complex  14640970   
254553             wildtype MshC His-tagged  17848100   

                       Organism UniprotID     ECNumber  parameter.startValue  \
110272        Rattus norvegicus  [P23457]  [1.1.1.213]              0.000025   
110291        Rattus norvegicus  [P23457]  [1.1.1.213]              0.000026   
110310        Rattus norvegicus  [P23457]  [1.1.1.213]              0.000022   
110332        Rattus norvegicus  [P23457]  [1.1.1.213]              0.000022   
200400        Bacillus subtilis        []  [3.4.21.62]              0.069300   
200406        Bacillus subtilis        []  [3.4.21.62]              0.051000   
254553  Mycobacterium smegmatis        []   [6.3.1.13]              0.005200   

        parameter.standardDeviation parameter.unit   pH  Temperature  
110272                 7.000000e-07              M  7.0         25.0  
110291                 1.100000e-06              M  7.0         25.0  
110310                 4.000000e-07              M  7.0         25.0  
110332                 7.000000e-07              M  7.0         25.0  
200400                 3.500000e-03              M  8.2         37.0  
200406                 2.600000e-03              M  8.2         37.0  
254553                 8.000000e-04              M  7.8         25.0

Clean up the units. Keep only `s^(-1)` and convert `mol*s^(-1)*mol^(-1)` and `NaN` to `s^(-1)`

In [47]:
mask = ((df['parameter.unit'].isnull()) | (df['parameter.unit'] == 'mol*s^(-1)*mol^(-1)'))
df.loc[mask, 'parameter.unit'] = 's^(-1)'

In [48]:
mask = df['parameter.unit'] == 's^(-1)'
df = df.loc[mask]

#### `pH`

In [49]:
np.sum(~df['pH'].isnull())

27334

#### `Temperature`

In [50]:
np.sum(~df['Temperature'].isnull())

25953

Most entries seem to have an associated pH and temperature.

# Save dataset

In [51]:
df

EntryID                                      Substrate  \
3             1                                 [H2O, Peptide]   
6            10        [(S)-Mandelate, Riboflavin-5-phosphate]   
8           100                     [Dihydrofolate, H+, NADPH]   
13          101                     [Dihydrofolate, H+, NADPH]   
18          102                     [Dihydrofolate, H+, NADPH]   
...         ...                                            ...   
333509     9874                 [H2O, N,N-Dimethylglycine, O2]   
333511     9875                 [H2O, N,N-Dimethylglycine, O2]   
333513     9876                 [H2O, N,N-Dimethylglycine, O2]   
333972     9970  [5alpha-Androstane-17beta-ol-3-one, H+, NADH]   
333980     9971   [5alpha-Androstane-3alpha,17beta-diol, NAD+]   

                                           EnzymeType  PubMedID  \
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)  12962477   
6                                     mutant MDH G81S  15311930   
8                                       wildtype DHFR  15740745   
13                                   mutant DHFR W30A  15740745   
18                                   mutant DHFR W30E  15740745   
...                                               ...       ...   
333509                                       wildtype  12912903   
333511                                   mutant H225Q  12912903   
333513                                   mutant Y259F  12912903   
333972                                  wildtype P26h  11141073   
333980                                  wildtype P26h  11141073   

                        Organism UniprotID     ECNumber  parameter.startValue  \
3                Bacillus lentus        []  [3.4.21.62]              220.0000   
6             Pseudomonas putida  [P20932]  [1.1.99.31]                2.8000   
8               Escherichia coli  [P0ABQ4]    [1.5.1.3]                6.1800   
13              Escherichia coli  [P0ABQ4]    [1.5.1.3]               14.5000   
18              Escherichia coli  [P0ABQ4]    [1.5.1.3]               11.5800   
...                          ...       ...          ...                   ...   
333509  Arthrobacter globiformis        []   [1.5.3.10]               10.6000   
333511  Arthrobacter globiformis        []   [1.5.3.10]                4.3500   
333513  Arthrobacter globiformis        []   [1.5.3.10]                0.0047   
333972      Mesocricetus auratus  [Q9Z245]  [1.1.1.184]                0.8500   
333980      Mesocricetus auratus  [Q9Z245]  [1.1.1.184]                1.2800   

        parameter.standardDeviation parameter.unit   pH  Temperature  
3                               NaN         s^(-1)  7.5         25.0  
6                               0.1         s^(-1)  7.5         20.0  
8                               NaN         s^(-1)  7.0         15.0  
13                              NaN         s^(-1)  7.0         15.0  
18                              NaN         s^(-1)  7.0         15.0  
...                             ...            ...  ...          ...  
333509                          NaN         s^(-1)  8.5         25.0  
333511                          NaN         s^(-1)  8.5         25.0  
333513                          NaN         s^(-1)  8.5         25.0  
333972                          NaN         s^(-1)  7.0         25.0  
333980                          NaN         s^(-1)  7.0         25.0  

[28129 rows x 12 columns]

In [56]:
fdir = '../../../../data/databases/Sabio-RK/'

In [53]:
df = df.reset_index(drop=True)

In [54]:
df.to_csv(fdir+'kcats.csv', index=False)

In [55]:
df.to_json(fdir+'kcats.json', index=False)